### ISSUE TITLE
Union Field: union_mode="smart" with default_factory and self reference ignores data that could be coerced

### ISSUE THREAD

**Issue by guillaumegenthial** [pydantic/pydantic#9101]
### Initial Checks

- [X] I confirm that I'm using Pydantic V2

### Description

The `union_mode` default `smart` gets confused if
- the attribute is a self referencing union
- the raw data needs to be coerced
- the Field has a default factory

In this situation, it ends up picking the default factory even if it should be capable of coercing the input.

NOTE: it seems that if you remove the default factory or don't use a self-referencing union, things work fine.

### Example Code

```Python
from typing import Union

from pydantic import BaseModel, Field


class Point(BaseModel):
    x: int
    y: int


class Points(BaseModel):
    # This works
    # points: Union[Point, "Points"] = Field(
    #     union_mode="smart",
    # )

    # This ignores the data
    points: Union[Point, "Points"] = Field(
        union_mode="smart",
        default_factory=lambda: Point(x=0, y=0),
    )


points = Points(
    points={
        "x": "1",
        "y": "1",
    },
)
print(points)  # points=Points(points=Point(x=0, y=0))


points = Points(
    points={
        "x": 1,
        "y": 1,
    },
)
print(points)  # points=Point(x=1, y=1)
```

### Python, Pydantic & OS Version

```Text
pydantic version: 2.6.4
        pydantic-core version: 2.16.3
          pydantic-core build: profile=release pgo=true
                 install path: .../envs/fragment-3.11/lib/python3.11/site-packages/pydantic
               python version: 3.11.7 | packaged by conda-forge | (main, Dec 23 2023, 14:38:07) [Clang 16.0.6 ]
                     platform: macOS-14.4-arm64-arm-64bit
             related packages: typing_extensions-4.9.0 mypy-1.8.0 fastapi-0.104.1 email-validator-2.1.0.post1
                       commit: unknown
```

### Related

https://github.com/pydantic/pydantic/issues/2135
https://github.com/pydantic/pydantic/pull/2092

**Comment by guillaumegenthial** [comment]
To fix, one option is to specify `union_mode=left_to_right` (see [Pydantic docs on Unions](https://docs.pydantic.dev/latest/concepts/unions/#union-modes)).

```python

class Points(BaseModel):
    # This works
    # points: Union[Point, "Points"] = Field(
    #     union_mode="smart",
    # )

    # This ignores the data
    points: Union[Point, "Points"] = Field(
        union_mode="left_to_right",
        default_factory=lambda: Point(x=0, y=0),
    )
```

**Comment by guillaumegenthial** [comment]
However, if you have a list of Unions, you can't use this trick

```python
from typing import Union

from pydantic import BaseModel, Field


class Point(BaseModel):
    x: int
    y: int


class Points(BaseModel):
    # This works
    # points: Union[Point, "Points"] = Field(
    #     union_mode="smart",
    # )

    # This ignores the data
    points: list[Union[Point, "Points"]] = Field(
        union_mode="left_to_right",
        default_factory=lambda: Point(x=0, y=0),
    )


points = Points(
    points=[
        {
            "x": "1",
            "y": "1",
        }
    ],
)
print(points)  # points=Points(points=Point(x=0, y=0))


points = Points(
    points=[
        {
            "x": 1,
            "y": 1,
        }
    ],
)
print(points)  # points=Point(x=1, y=1)
```

which raises

```bash
TypeError: The following constraints cannot be applied to list[typing.Union[__main__.Point, __main__.Points]]: 'union_mode'
```

**Comment by sydney-runkle** [comment]
> However, if you have a list of Unions, you can't use this trick

You can use `Annotated` on the `Union` within the List in order to apply this constraint via the `Field(union_mode='left_to_right')` annotation.

Given that, I think this issue is resolved...

**Comment by guillaumegenthial** [comment]
@sydney-runkle thanks for the quick reply.

Indeed this works

```python
class Points(BaseModel):
    # This works
    points: list[
        Annotated[Union[Point, "Points"], Field(union_mode="left_to_right")]
    ] = Field(default_factory=list)

```

**Comment by guillaumegenthial** [comment]
> Given that, I think this issue is resolved...

I would argue that it's not quite resolved. There is a _workaround_ though.

Why ?
1. The behavior is not consistent (adding a `default_factory` should not change the resolved type ?)
2. It's unexpected (data that would otherwise be coerced is ignored)

This caused major issues on our product FYI and was quite hard to track down...

**Comment by sydney-runkle** [comment]
@guillaumegenthial,

Ah yes you're right. I didn't understand the issue in its entirety! Thanks for following up.

**Comment by sydney-runkle** [comment]
Fixed now with `pydantic-core` v2.22.0, I believe via https://github.com/pydantic/pydantic-core/pull/1398.

New behavior:

```py
from typing import Union

from pydantic import BaseModel, Field


class Point(BaseModel):
    x: int
    y: int


class Points(BaseModel):
    # This works
    # points: Union[Point, "Points"] = Field(
    #     union_mode="smart",
    # )

    # This ignores the data
    points: Union[Point, "Points"] = Field(
        union_mode="smart",
        default_factory=lambda: Point(x=0, y=0),
    )


points = Points(
    points={
        "x": "1",
        "y": "1",
    },
)
print(points)  # points=Point(x=1, y=1)


points = Points(
    points={
        "x": 1,
        "y": 1,
    },
)
print(points)  # points=Point(x=1, y=1)
```

**Comment by sydney-runkle** [comment]
Fix should be released with v2.9!
